ext/soap/soap.c

```diff
@@ -584,23 +590,26 @@ PHP_METHOD(SoapFault, __construct)
 {
 	char *fault_string = NULL, *fault_code = NULL, *fault_actor = NULL, *name = NULL, *fault_code_ns = NULL;
 	size_t fault_string_len, fault_actor_len = 0, name_len = 0, fault_code_len = 0;
-	zval *code = NULL, *details = NULL, *headerfault = NULL, *this_ptr;
-
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "zs|s!z!s!z!",
-		&code,
-		&fault_string, &fault_string_len,
-		&fault_actor, &fault_actor_len,
-		&details, &name, &name_len, &headerfault) == FAILURE) {
-		RETURN_THROWS();
-	}
-
-	if (Z_TYPE_P(code) == IS_NULL) {
-	} else if (Z_TYPE_P(code) == IS_STRING) {
-		fault_code = Z_STRVAL_P(code);
-		fault_code_len = Z_STRLEN_P(code);
-	} else if (Z_TYPE_P(code) == IS_ARRAY && zend_hash_num_elements(Z_ARRVAL_P(code)) == 2) {
-		zval *t_ns = zend_hash_index_find(Z_ARRVAL_P(code), 0);
-		zval *t_code = zend_hash_index_find(Z_ARRVAL_P(code), 1);
+	zval *details = NULL, *headerfault = NULL, *this_ptr;
+	zend_string *code_str;
+	HashTable *code_ht;
+
+	ZEND_PARSE_PARAMETERS_START(2, 6)
+		Z_PARAM_STR_OR_ARRAY_HT_OR_NULL(code_str, code_ht)
+		Z_PARAM_STRING(fault_string, fault_string_len)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STRING_OR_NULL(fault_actor, fault_actor_len)
+		Z_PARAM_ZVAL_OR_NULL(details)
+		Z_PARAM_STRING_OR_NULL(name, name_len)
+		Z_PARAM_ZVAL_OR_NULL(headerfault)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (code_str) {
+		fault_code = ZSTR_VAL(code_str);
+		fault_code_len = ZSTR_LEN(code_str);
+	} else if (code_ht && zend_hash_num_elements(code_ht) == 2) {
```

Comment:
I'd move the `zend_hash_num_elements(code_ht) == 2` condition into the if inside. We can check all "Invalid fault code" conditions at the same time.

Reply:
Hopefully, I managed to find the correct condition

Reply:
With your latest change, moving the zend_hash_num_elements(code_ht) == 2 into a separate condition is not actually needed anymore (though it doesn't matter either way).

Reply:
I got rid of nesting at last.